File: backend/models/reservation.py
from typing import Optional, Dict, Any, List
from datetime import date, datetime
from database import db
from controllers.settings_controller import SettingsController

class EmployeeReservation:
    def __init__(self, **kwargs):
        self.id = kwargs.get('id')
        self.employee_id = kwargs.get('employee_id')
        self.start_date = kwargs.get('start_date')
        self.end_date = kwargs.get('end_date')
        self.reserved_hours_per_day = kwargs.get('reserved_hours_per_day', 0)
        self.reason = kwargs.get('reason')
        self.status = kwargs.get('status', 'active')
        self.created_at = kwargs.get('created_at')
        self.updated_at = kwargs.get('updated_at')
    
    @staticmethod
    def create(employee_id: int, start_date: date, end_date: date, 
               reserved_hours_per_day: float, reason: str = None) -> 'EmployeeReservation':
        """Create a new reservation"""
        if end_date < start_date:
            raise ValueError("End date must be after or equal to start date")
        
        max_hours = SettingsController.get_work_hours_per_day()
        if reserved_hours_per_day < 0 or reserved_hours_per_day > max_hours:
            raise ValueError(f"Reserved hours per day must be between 0 and {max_hours}")
        
        # Check for overlapping reservations
        overlap_query = '''
            SELECT id FROM employee_reservations
            WHERE employee_id = ? 
                AND status = 'active'
                AND (
                    (start_date <= ? AND end_date >= ?)
                    OR (start_date <= ? AND end_date >= ?)
                    OR (start_date >= ? AND end_date <= ?)
                )
        '''
        existing = db.fetch_one(overlap_query, (
            employee_id,
            start_date, start_date,
            end_date, end_date,
            start_date, end_date
        ))
        
        if existing:
            raise ValueError("Overlapping reservation already exists for this period")
        
        query = '''
            INSERT INTO employee_reservations 
            (employee_id, start_date, end_date, reserved_hours_per_day, reason)
            VALUES (?, ?, ?, ?, ?)
        '''
        cursor = db.execute(query, (employee_id, start_date, end_date, 
                                   reserved_hours_per_day, reason))
        db.commit()
        return EmployeeReservation.get_by_id(cursor.lastrowid)
    
    @staticmethod
    def get_by_id(reservation_id: int) -> Optional['EmployeeReservation']:
        query = 'SELECT * FROM employee_reservations WHERE id = ?'
        row = db.fetch_one(query, (reservation_id,))
        return EmployeeReservation(**row) if row else None
    
    @staticmethod
    def get_by_employee(employee_id: int, include_cancelled: bool = False) -> List['EmployeeReservation']:
        """Get all reservations for an employee"""
        if include_cancelled:
            query = '''
                SELECT * FROM employee_reservations 
                WHERE employee_id = ?
                ORDER BY start_date DESC
            '''
        else:
            query = '''
                SELECT * FROM employee_reservations 
                WHERE employee_id = ? AND status = 'active'
                ORDER BY start_date DESC
            '''
        rows = db.fetch_all(query, (employee_id,))
        return [EmployeeReservation(**row) for row in rows]
    
    @staticmethod
    def get_active_reservations_for_date_range(employee_id: int, start_date: date, 
                                                end_date: date) -> List['EmployeeReservation']:
        """Get active reservations that overlap with a given date range"""
        query = '''
            SELECT * FROM employee_reservations
            WHERE employee_id = ? 
                AND status = 'active'
                AND (
                    (start_date <= ? AND end_date >= ?)
                    OR (start_date <= ? AND end_date >= ?)
                    OR (start_date >= ? AND end_date <= ?)
                )
            ORDER BY start_date
        '''
        rows = db.fetch_all(query, (
            employee_id,
            start_date, start_date,
            end_date, end_date,
            start_date, end_date
        ))
        return [EmployeeReservation(**row) for row in rows]
    
    def update(self, **kwargs) -> 'EmployeeReservation':
        """Update reservation details"""
        updates = []
        params = []
        
        if 'start_date' in kwargs:
            updates.append('start_date = ?')
            params.append(kwargs['start_date'])
        if 'end_date' in kwargs:
            updates.append('end_date = ?')
            params.append(kwargs['end_date'])
        if 'reserved_hours_per_day' in kwargs:
            max_hours = SettingsController.get_work_hours_per_day()
            if kwargs['reserved_hours_per_day'] < 0 or kwargs['reserved_hours_per_day'] > max_hours:
                raise ValueError(f"Reserved hours per day must be between 0 and {max_hours}")
            updates.append('reserved_hours_per_day = ?')
            params.append(kwargs['reserved_hours_per_day'])
        if 'reason' in kwargs:
            updates.append('reason = ?')
            params.append(kwargs['reason'])
        if 'status' in kwargs:
            updates.append('status = ?')
            params.append(kwargs['status'])
        
        if updates:
            updates.append('updated_at = CURRENT_TIMESTAMP')
            params.append(self.id)
            query = f'UPDATE employee_reservations SET {", ".join(updates)} WHERE id = ?'
            db.execute(query, tuple(params))
            db.commit()
        
        return EmployeeReservation.get_by_id(self.id)
    
    def cancel(self) -> 'EmployeeReservation':
        """Cancel this reservation"""
        return self.update(status='cancelled')
    
    def delete(self) -> bool:
        """Delete this reservation"""
        try:
            query = 'DELETE FROM employee_reservations WHERE id = ?'
            db.execute(query, (self.id,))
            db.commit()
            return True
        except Exception as e:
            db.rollback()
            raise e
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            'id': self.id,
            'employee_id': self.employee_id,
            'start_date': self.start_date.isoformat() if isinstance(self.start_date, date) else self.start_date,
            'end_date': self.end_date.isoformat() if isinstance(self.end_date, date) else self.end_date,
            'reserved_hours_per_day': self.reserved_hours_per_day,
            'reason': self.reason,
            'status': self.status,
            'created_at': self.created_at,
            'updated_at': self.updated_at
        }

File: backend-node/src/models/reservation.js
const db = require('../database');
const SettingsController = require('../controllers/settingsController');

class EmployeeReservation {
  constructor(data) {
    this.id = data.id;
    this.employee_id = data.employee_id;
    this.start_date = data.start_date;
    this.end_date = data.end_date;
    this.reserved_hours_per_day = data.reserved_hours_per_day || 0;
    this.reason = data.reason;
    this.status = data.status || 'active';
    this.created_at = data.created_at;
    this.updated_at = data.updated_at;
  }

  static create(employeeId, startDate, endDate, reservedHoursPerDay, reason = null) {
    if (endDate < startDate) {
      throw new Error("End date must be after or equal to start date");
    }

    const maxHours = SettingsController.getWorkHoursPerDay();
    if (reservedHoursPerDay < 0 || reservedHoursPerDay > maxHours) {
      throw new Error(`Reserved hours per day must be between 0 and ${maxHours}`);
    }

    const result = db.execute(
      `INSERT INTO employee_reservations
       (employee_id, start_date, end_date, reserved_hours_per_day, reason)
       VALUES (?, ?, ?, ?, ?)`,
      [employeeId, startDate, endDate, reservedHoursPerDay, reason]
    );

    return EmployeeReservation.getById(result.lastInsertRowid);
  }

  static getById(reservationId) {
    const row = db.fetchOne('SELECT * FROM employee_reservations WHERE id = ?', [reservationId]);
    return row ? new EmployeeReservation(row) : null;
  }

  static getByEmployee(employeeId, includeCancelled = false) {
    let query;
    if (includeCancelled) {
      query = `SELECT * FROM employee_reservations
               WHERE employee_id = ?
               ORDER BY start_date DESC`;
    } else {
      query = `SELECT * FROM employee_reservations
               WHERE employee_id = ? AND status = 'active'
               ORDER BY start_date DESC`;
    }
    const rows = db.fetchAll(query, [employeeId]);
    return rows.map(row => new EmployeeReservation(row));
  }

  static getActiveReservationsForDateRange(employeeId, startDate, endDate) {
    // Use strict inequalities to allow adjacent date ranges
    // Two ranges overlap if: start1 < end2 AND end1 > start2
    const rows = db.fetchAll(`
      SELECT * FROM employee_reservations
      WHERE employee_id = ?
        AND status = 'active'
        AND start_date < ?
        AND end_date > ?
      ORDER BY start_date
    `, [employeeId, endDate, startDate]);

    return rows.map(row => new EmployeeReservation(row));
  }

  update(data) {
    const updates = [];
    const params = [];

    if (data.start_date !== undefined) {
      updates.push('start_date = ?');
      params.push(data.start_date);
    }
    if (data.end_date !== undefined) {
      updates.push('end_date = ?');
      params.push(data.end_date);
    }
    if (data.reserved_hours_per_day !== undefined) {
      const maxHours = SettingsController.getWorkHoursPerDay();
      if (data.reserved_hours_per_day < 0 || data.reserved_hours_per_day > maxHours) {
        throw new Error(`Reserved hours per day must be between 0 and ${maxHours}`);
      }
      updates.push('reserved_hours_per_day = ?');
      params.push(data.reserved_hours_per_day);
    }
    if (data.reason !== undefined) {
      updates.push('reason = ?');
      params.push(data.reason);
    }
    if (data.status !== undefined) {
      updates.push('status = ?');
      params.push(data.status);
    }

    if (updates.length > 0) {
      updates.push('updated_at = CURRENT_TIMESTAMP');
      params.push(this.id);
      db.execute(`UPDATE employee_reservations SET ${updates.join(', ')} WHERE id = ?`, params);
    }

    return EmployeeReservation.getById(this.id);
  }

  cancel() {
    return this.update({ status: 'cancelled' });
  }

  delete() {
    db.execute('DELETE FROM employee_reservations WHERE id = ?', [this.id]);
    return true;
  }

  toDict() {
    const formatDate = (d) => {
      if (d === null || d === undefined) return null;
      if (typeof d === 'string') return d;
      if (d instanceof Date) return d.toISOString().split('T')[0];
      return d;
    };

    return {
      id: this.id,
      employee_id: this.employee_id,
      start_date: formatDate(this.start_date),
      end_date: formatDate(this.end_date),
      reserved_hours_per_day: this.reserved_hours_per_day,
      reason: this.reason,
      status: this.status,
      created_at: this.created_at,
      updated_at: this.updated_at
    };
  }
}

module.exports = EmployeeReservation;
